Upgrade cirrus test framework

liuchengxu opened the pull request:
This PR is primarily to fix some issues in the current test setup I found while working on another branch, e.g., the slot notification is not enabled when starting an executor. Only some minor refactoring to the production code, all the other changes are for tests.

Now, `subspace-test-service` can be used for running a pure primary node with farmer function, `run_executor` in `cirrus-test-service` behaves similar to in the production code (subspace-node.rs), which starts a primary chain full node as well as a secondary chain. There is still some room for improvement, but it's enough for me for the testing purpose for now.

This PR is expected to also fix the flaky test https://github.com/subspace/subspace/blob/660db097e8/cumulus/client/cirrus-executor/src/tests.rs#L167 .

Close #505

liuchengxu committed:
Rename SubspaceTestNode to PrimaryTestNode

Trying to reflect the primary and secondary chain design.

liuchengxu committed:
Inline `new_full` in subspace-test-service

liuchengxu committed:
Remove unused code and some renamings

liuchengxu committed:
Expose more components in cirrus NewFull

We'll reuse `cirrus::node::service::new_full` in the following commit,
these exposed components are necessary for the test usage.

liuchengxu committed:
Make RuntimeApi and ExecutorDispatch generic in cirrus new_full

liuchengxu committed:
Expose executor in cirrus `new_full`

liuchengxu committed:
Reuse cirrus_node::service::new_full in test

liuchengxu committed:
Some minor cleanups

liuchengxu committed:
Always enable the slot notification when running an executor

liuchengxu committed:
Remove unused new_partial in cirrus-test-service

liuchengxu committed:
Hardcode Alice as the dev seed in secondary chain configuration

Now the bundle production in tests works again no matter what initial key is passed to the TestNode constructor.

liuchengxu committed:
Mock the bundles processing manually

I can reproduce the test failure with a busy CPU locally, this patch should
fix the occasional test failure.

liuchengxu committed:
Some naming and comment cleanups

liuchengxu committed:
Rename `start_secondary_node` to `run_executor`

nazar-pc commented on cumulus/node/src/service.rs:
```diff
@@ -261,48 +275,48 @@ where
+		});
+	spawn_essential.spawn_essential_blocking("cirrus-gossip", None, Box::pin(executor_gossip));
+
+	let new_full = NewFull {
```
Why not as part of `NewFull` like for everything else?

liuchengxu replied:
I think the initial reason is not to make `NewFull` too complicated and deal with the compiling from the very beginning, but it can be solved much easier now, updated.

nazar-pc commented on cumulus/test/service/src/lib.rs:
```diff
@@ -400,7 +400,8 @@ pub fn node_config(
 		disable_grandpa: false,
-		dev_key_seed: Some(key_seed),
+		// Force using Alice as the dev seed as the specified executor authority in primary chain spec is Alice.
+		dev_key_seed: Some(Keyring::Alice.to_seed()),
```
This doesn't seem right. You'll have to change it back in order to support multiple identities, why not specifying correct `key_seed` from the outside instead?

liuchengxu replied:
I was hesitant/lazy to fix the node name and dev key seed properly :P, it works for now, but you're right, we'll have to fix it in the future anyway, updated in https://github.com/subspace/subspace/pull/523/commits/1057adc3131b433c9e63edb9a623d4751a11fbe1.

Also rebased on the latest main, only a few minor conflicts.

nazar-pc commented on cumulus/client/cirrus-executor/src/tests.rs:
```diff
@@ -133,8 +136,35 @@ async fn execution_proof_creation_and_verification_should_work() {
+	charlie
+		.executor
+		.clone()
+		.process_bundles(
```
This was a legacy method we wanted to remove, would be preferable to not add more uses of it and instead observe primary chain or something like that.

liuchengxu replied:
Yeah, I'd love to but didn't find a nice way that just makes use of the native procedure but still ensures the test txs can be included in some block. It's indefinite if you have a really busy CPU from my local experiment.

liuchengxu committed:
Make executor part of cirrus NewFull instead of returning a tuple

liuchengxu committed:
Handle the node name and dev_key_seed properly

New notation:

- Ferdie(last item in enum Keyring): pure primary node with farmer function.
- Alice(first in enum Keyring): executor authority node.
- Bob(second in enum Keyring): executor full node.

More executors in the future will be added following the order defined in enum
Keyring(Charlie,...).

Most of the changes in this commit are about renamings:

        Before  =>  After
         alice  =>  ferdie
       charlie  =>  alice
          dave  =>  bob
alice_executor  =>  alice

nazar-pc reviewed (dismissed):
Looks fine, but clippy needs to be fixed

liuchengxu committed:
Make clippy happy
